src/main/java/com/google/devtools/build/lib/skyframe/PackageArgsFunction.java

```diff
@@ -0,0 +1,124 @@
+import com.google.devtools.build.skyframe.SkyValue;
+import java.util.Map;
+import javax.annotation.Nullable;
+import net.starlark.java.eval.EvalException;
+
+/** A @{link SkyFunction} that returns the {@link PackageArgs} for a given repository. */
+public class PackageArgsFunction implements SkyFunction {
+  public static final PackageArgsFunction INSTANCE = new PackageArgsFunction();
+
+  /** {@link SkyValue} wrapping a PackageArgs. */
+  public static final class PackageArgsValue implements SkyValue {
+    public static final PackageArgsValue EMPTY = new PackageArgsValue(PackageArgs.EMPTY);
+
+    private final PackageArgs packageArgs;
+
+    public PackageArgsValue(PackageArgs packageArgs) {
+      this.packageArgs = packageArgs;
+    }
+
+    public PackageArgs getPackageArgs() {
+      return packageArgs;
+    }
+
+    @Override
+    public int hashCode() {
+      return packageArgs.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object other) {
+      if (other instanceof PackageArgsValue that) {
+        return that.packageArgs.equals(packageArgs);
+      } else {
+        return false;
+      }
+    }
+  }
+
+  public static Key key(RepositoryName repoName) {
+    return new Key(repoName);
+  }
+
+  /** Key type for {@link RepoFileValue}. */
+  public static class Key extends AbstractSkyKey<RepositoryName> {
+
+    private Key(RepositoryName repoName) {
+      super(repoName);
+    }
+
+    @Override
+    public SkyFunctionName functionName() {
+      return SkyFunctions.PACKAGE_ARGS;
+    }
+  }
+
+  private PackageArgsFunction() {}
+
+  @Nullable
+  @Override
+  public SkyValue compute(SkyKey skyKey, Environment env)
+      throws SkyFunctionException, InterruptedException {
+    RepositoryName repositoryName = (RepositoryName) skyKey.argument();
+
+    RepoFileValue repoFileValue = (RepoFileValue) env.getValue(RepoFileValue.key(repositoryName));
+    RepositoryMappingValue repositoryMappingValue =
+        (RepositoryMappingValue) env.getValue(RepositoryMappingValue.key(repositoryName));
+
+    if (env.valuesMissing()) {
+      return null;
+    }
+    String repoDisplayName = RepoFileFunction.getDisplayNameForRepo(repositoryName, null);
+
+    PackageArgs.Builder pkgArgsBuilder = PackageArgs.builder();
+    LabelConverter labelConverter =
+        new LabelConverter(
+            PackageIdentifier.create(repositoryName, PathFragment.EMPTY_FRAGMENT),
+            repositoryMappingValue.getRepositoryMapping());
+    try {
+      for (Map.Entry<String, Object> kwarg : repoFileValue.packageArgsMap().entrySet()) {
+        PackageArgs.processParam(
+            kwarg.getKey(),
+            kwarg.getValue(),
+            "repo() argument '" + kwarg.getKey() + "'",
+            labelConverter,
+            pkgArgsBuilder);
+      }
+    } catch (EvalException e) {
+      env.getListener().handle(Event.error(e.getMessageWithStack()));
+      throw new RepoFileFunctionException(
```

Comment:
let's stick with the idiom where we create an exception type for this skyfunction specifically, and keep RepoFileFunctionException ctors private?

Reply:
Done. It required more boilerplate than I'd like, but what can I do, this is Java...